ENH: Add all font metrics for base 14 Type 1 PDF fonts.

PJBrs opened the pull request:
This patch includes font metrics for the standard 14 fonts. This is intended to be useful for generating a text appearance stream, especially if you want to take into account right-aligned or centred text. (I have some other patches that include this as well as text wrapping.)

Note that some of this information was already included in _font_widths.py, but that information is incomplete. I thought it better to copy this information from pdfminer.six and be able to potentially benefit from their work later on, than to improve on what already was included here.

The first three patches introduce the new functionality. The last three patches are for moving the Font class to the new font metrics information and removing the old _font_widths.py file.

This is what the spec has to say about it:

> 9.6.2.2 Standard Type 1 fonts (standard 14 fonts) (PDF 1.0-1.7)
> 
> The PostScript language names of 14 Type 1 fonts, known as the standard 14 fonts, are as follows:
> Times-Roman, Helvetica, Courier, Symbol, Times-Bold, Helvetica-Bold, Courier-Bold, ZapfDingbats,
> Times-Italic, Helvetica-Oblique, Courier-Oblique, Times-BoldItalic, Helvetica-BoldOblique, CourierBoldOblique.
> 
> In PDF 1.0 to PDF 1.7, the FirstChar, LastChar, Widths and FontDescriptor (see Table 109: Entries in
> a Type 1 font dictionary") were optional in Type 1 font dictionaries for the standard 14 fonts. PDF
> processors supporting PDF 1.0 to PDF 1.7 files shall have these fonts, or their font metrics and suitable
> substitution fonts, available.
> 
> These fonts, or their font metrics and suitable substitution fonts, shall be available to the PDF processor.

PJBrs commented:
Weird, some tests fail that seem to run fine locally...

PJBrs commented:
```
pytest3 tests/test_workflows.py
=========================================================== test session starts ===========================================================
platform linux -- Python 3.9.23, pytest-8.4.0, pluggy-1.6.0
rootdir: /home/pj/Src/python3-pypdf/pypdf
configfile: pyproject.toml
plugins: socket-0.7.0, timeout-2.1.0
collected 107 items                                                                                                                       

tests/test_workflows.py ........................................................................................................... [100%]

========================================================== 107 passed in 31.44s ===========================================================
```

stefan6419846 commented on pypdf/_codecs/base14_fontmetrics.py:
```diff
@@ -0,0 +1,4464 @@
+                    props["Flags"] = 64
+            elif k == "FontBBox":
+                props[k] = tuple(map(float, f[1:5]))
+        print("# -*- python -*-")
```
This does not seem to match the actual file content. Has this function really been used to generate the remaining content?

PJBrs replied:
I'll check it myself.

I just copied the file verbatim from pdfminer.six

PJBrs replied:
https://github.com/pdfminer/pdfminer.six/blob/master/pdfminer/fontmetrics.py

PJBrs replied:
Hmmm... You're right, this is what I get for "hv.afm":

```
# -*- python -*-
FONT_METRICS = {
 'Helvetica': ({'FontName': 'Helvetica', 'Flags': 0, 'FontFamily': 'Helvetica', 'FontWeight': 'Medium', 'ItalicAngle': 0.0, 'FontBBox': (-166.0, -225.0, 1000.0, 931.0), 'CapHeight': 718.0, 'XHeight': 523.0, 'Ascent': 718.0, 'Descent': -207.0}, {32: 278, 33: 278, 34: 355, 35: 556, 36: 556, 37: 889, 38: 667, 39: 222, 40: 333, 41: 333, 42: 389, 43: 584, 44: 278, 45: 333, 46: 278, 47: 278, 48: 556, 49: 556, 50: 556, 51: 556, 52: 556, 53: 556, 54: 556, 55: 556, 56: 556, 57: 556, 58: 278, 59: 278, 60: 584, 61: 584, 62: 584, 63: 556, 64: 1015, 65: 667, 66: 667, 67: 722, 68: 722, 69: 667, 70: 611, 71: 778, 72: 722, 73: 278, 74: 500, 75: 667, 76: 556, 77: 833, 78: 722, 79: 778, 80: 667, 81: 778, 82: 722, 83: 667, 84: 611, 85: 722, 86: 667, 87: 944, 88: 667, 89: 667, 90: 611, 91: 278, 92: 278, 93: 278, 94: 469, 95: 556, 96: 222, 97: 556, 98: 556, 99: 500, 100: 556, 101: 556, 102: 278, 103: 556, 104: 556, 105: 222, 106: 222, 107: 500, 108: 222, 109: 833, 110: 556, 111: 556, 112: 556, 113: 556, 114: 333, 115: 500, 116: 278, 117: 556, 118: 500, 119: 722, 120: 500, 121: 500, 122: 500, 123: 334, 124: 260, 125: 334, 126: 584, 161: 333, 162: 556, 163: 556, 164: 167, 165: 556, 166: 556, 167: 556, 168: 556, 169: 191, 170: 333, 171: 556, 172: 333, 173: 333, 174: 500, 175: 500, 177: 556, 178: 556, 179: 556, 180: 278, 182: 537, 183: 350, 184: 222, 185: 333, 186: 333, 187: 556, 188: 1000, 189: 1000, 191: 611, 193: 333, 194: 333, 195: 333, 196: 333, 197: 333, 198: 333, 199: 333, 200: 333, 202: 333, 203: 333, 205: 333, 206: 333, 207: 333, 208: 1000, 225: 1000, 227: 370, 232: 556, 233: 778, 234: 1000, 235: 365, 241: 889, 245: 278, 248: 222, 249: 611, 250: 944, 251: 611}),
}
```

stefan6419846 replied:
Ideally, we can keep this function and just adapt it to output the desired format correctly. In your case, it seems like it might be sufficient to sort `fonts` by key before outputting it.

PJBrs replied:
"besides converting the CIDs to characters."

I don't know how to do this, unfortunately... I might be able to learn.

stefan6419846 replied:
No, when generating the code for the fonts, we should download the archive file automatically instead of asking the developer to do so manually.

The CTAN file has more fonts than the GitHub repository and the original version. Thus I would probably prefer to either use http://web.archive.org/web/20110531171921/http://www.adobe.com/content/dam/Adobe/en/devnet/font/pdfs/Core14_AFMs.zip (the archived copy) or the current link https://download.macromedia.com/pub/developer/opentype/tech-notes/Core14_AFMs.zip which seems to serve the same files.

stefan6419846 commented on pypdf/_codecs/base14_fontmetrics.py:
```diff
@@ -0,0 +1,4464 @@
+
+"""
+
+###  BEGIN Verbatim copy of the license part
```
This does not match the requirements on copyright notices (see our other files with third-party content as well). Additionally, you mention that this is somehow related to *pdfminer.six*. Which parts are retrieved from there and where is the corresponding notice?

PJBrs replied:
The whole file was copied from them, including this notice. I'll try and clarify where possible!

PJBrs replied:
I think I sufficiently clarified the license situation with the first patch in this PR.

stefan6419846 commented on pypdf/_codecs/base14_fontmetrics.py:
```diff
@@ -0,0 +1,4464 @@
+
+
+FONT_METRICS : Dict[str, Tuple[Dict[str, object], Dict[Any, float]]]  = {
+    "Courier": (
```
Instead of doing this complex typing and unclear indexing when using the data, I would suggest to use actual data containers for each font.

Example container:

```python
@dataclass(frozen=True)
class Font:
    name: str
    family: str
    weight: str

    ascent: float
    descent: float
    cap_height: float
    x_height: float
    italic_angle: float
    flags: int
    bbox: Tuple[float, float, float, float]

    character_widths: Dict[str, int]
```

PJBrs replied:
This is roughly what we have over here: https://github.com/py-pdf/pypdf/blob/main/pypdf/_text_extraction/_layout_mode/_font.py

Do I try to replace this? Or do I add a separate font.py in pypdf?

PJBrs replied:
OK. Jus for the record - I'm mostly a novice with coding, so I'm quite out of my depth here. What you're saying is that, although we _can_ add that big file with font metrics, ideally, we should not use it directly, but indirectly via a dedicated class that parses its content. Typically, this would be a dataclass, if I understand correctly, or it would be a dedicated font class, that could be instantiated with something like:
helvetica = Font("Helvetica") 

(Can something like this also be done with a dataclass?)

Questions I'd still have - I don't know how specific all these metrics that I'd like to add are for Type 1 fonts, and I know very little about fonts in general. So, I'd probably be able to make a class for _this_ data, but I don't know whether it would work for OTF / TTF fonts. ... But maybe that's what you mean? Adding a dedicated
@dataclass
class CoreFont() ?

PJBrs replied:
I've been looking at pdfminer.six's PDFFont class, https://github.com/pdfminer/pdfminer.six/blob/51683b2528e2aa685dd8b9e61f6ccf9f76a59a62/pdfminer/pdffont.py#L869 , and the more I think about it, the current Font class used by pypdf for text extraction is actually a specific case of what could be a more generic font class.

stefan6419846 commented:
Thanks for the further changes. There still are some further changes I would like to see:

* Move the generator script code into *pypdf* itself instead of the resources. Otherwise, linting does not trigger for example.
* To which extent does the generator script re-use code from *pdfminer.six*?
* Why is there such a "strange" copyright detection logic instead of just reading the corresponding fields?
* Although there might be changes in a later PR, I would still prefer to have a proper object-oriented approach for the new data. Past has shown that nothing is more permanent than a temporary solution.

PJBrs commented:
> Thanks for the further changes. There still are some further changes I would like to see:
> 
> * Move the generator script code into _pypdf_ itself instead of the resources. Otherwise, linting does not trigger for example.

No problem!

> * To which extent does the generator script re-use code from _pdfminer.six_?

I think about 20%? The resulting formatting is still very like pdfminer (although their original script did not produce unicode codes but ints). 

> * Why is there such a "strange" copyright detection logic instead of just reading the corresponding fields?

This is how it is in the original AFM files:

`Comment Copyright (c) 1985, 1987, 1989, 1990, 1997 Adobe Systems Incorporated.  All Rights Reserved.
`

and 

`Notice Copyright (c) 1985, 1987, 1989, 1990, 1997 Adobe Systems Incorporated.  All Rights Reserved.Helvetica is a trademark of Linotype-Hell AG and/or its subsidiaries.
`

So, the same copyright is in two locations, one after Comment and one after Notice. Only the second, however, includes mention of the trademark. And do notice the lack of a space between "Reserved." and "Helvetica". So, with all that, you indeed get a rather strange copyright detection logic!

> * Although there might be changes in a later PR, I would still prefer to have a proper object-oriented approach for the new data. Past has shown that nothing is more permanent than a temporary solution.

OK. Then I think I need more information about what that would entail. What I _think_ it does:

- Keep all information as it is now
- For the widths information, this can be used as I propose using it now, in `_cmap.build_font_width_map`
- Add a `@dataclass font_descriptor` to _cmap.py that can initialise itself using the information in core_fontmetrics.py
- Ideally, such a dataclass could actually use either the Type 1 font information from core_fontmetrics, or an existing /Descriptors dict within an embedded /Font resource. But I would need an easy test solution for that case (ideally, a form that doesn't use one of the core fonts).

Is that correct? 

If not, perhaps it might be easier to reach out via discord, where I use the same user name.

stefan6419846 commented:
In the first step, it would be sufficient for me to just create a new dataclass with the properties outlined previously. Merging it with the existing class can always happen later, but I would like to see real structured data for now. We basically need to define the new dataclass and adapt the generator script to call the dataclass constructor accordingly.

Regarding the *pdfminer.six* code and the copyrights, I will need to have another look at it in the next days.

PJBrs commented:
I'm afraid that means I'm officially out of my depth, I think I just lack the prerequisite knowledge to understand what you intend the generator script to do. If possible, please contact me on Discord one of these days so that we can discuss.

And as always, thanks for your comments!

PJBrs commented:
P.S., this is what Google Gemini thinks:

```
import re
from dataclasses import dataclass
from typing import Dict, Tuple

@dataclass(frozen=True)
class Font:
    """
    Dataclass to store parsed information from a Type 1 font AFM file.
    """
    name: str
    family: str
    weight: str

    ascent: float
    descent: float
    cap_height: float
    x_height: float
    italic_angle: float
    flags: int
    bbox: Tuple[float, float, float, float]

    character_widths: Dict[str, int]

def parse_afm_file(afm_content: str) -> Font:
    """
    Parses the content of a Type 1 font AFM file and returns a Font dataclass instance.

    Args:
        afm_content: A string containing the full content of the AFM file.

    Returns:
        A Font dataclass instance populated with the parsed data.

    Raises:
        ValueError: If essential font properties are missing or malformed in the AFM content.
    """
    font_properties = {}
    character_widths = {}
    in_char_metrics_section = False

    lines = afm_content.splitlines()

    for line in lines:
        line = line.strip()
        if not line:
            continue

        # Check for start/end of character metrics section
        if line == "StartCharMetrics":
            in_char_metrics_section = True
            continue
        elif line == "EndCharMetrics":
            in_char_metrics_section = False
            continue

        if in_char_metrics_section:
            # Parse character metrics: C <char_code>; WX <width_x>; N <char_name>; ...
            match = re.match(r"C\s+\d+;\s+WX\s+([\d.]+);\s+N\s+([a-zA-Z0-9_.]+);", line)
            if match:
                width = int(float(match.group(1))) # Widths are typically integers in AFM
                char_name = match.group(2)
                character_widths[char_name] = width
        else:
            # Parse general font properties
            parts = line.split(' ', 1) # Split only on the first space
            if len(parts) == 2:
                key, value = parts[0], parts[1]
                font_properties[key] = value

    # Extract and convert properties, handling potential missing values
    try:
        name = font_properties.get("FontName", "Unknown")
        family = font_properties.get("FamilyName", "Unknown")
        weight = font_properties.get("Weight", "Unknown")

        ascent = float(font_properties.get("Ascender", 0.0))
        descent = float(font_properties.get("Descender", 0.0))
        cap_height = float(font_properties.get("CapHeight", 0.0))
        x_height = float(font_properties.get("XHeight", 0.0))
        italic_angle = float(font_properties.get("ItalicAngle", 0.0))

        # Calculate flags: bit 0 is set if IsFixedPitch is true
        is_fixed_pitch = font_properties.get("IsFixedPitch", "false").lower() == "true"
        flags = 1 if is_fixed_pitch else 0

        # Parse FontBBox
        bbox_str = font_properties.get("FontBBox", "0 0 0 0")
        bbox_values = tuple(map(float, bbox_str.split()))
        if len(bbox_values) != 4:
            raise ValueError(f"Malformed FontBBox: {bbox_str}")
        bbox = bbox_values

    except (KeyError, ValueError) as e:
        raise ValueError(f"Error parsing AFM file: Missing or malformed property - {e}")

    return Font(
        name=name,
        family=family,
        weight=weight,
        ascent=ascent,
        descent=descent,
        cap_height=cap_height,
        x_height=x_height,
        italic_angle=italic_angle,
        flags=flags,
        bbox=bbox,
        character_widths=character_widths
    )
```

stefan6419846 commented:
I have just used your script and adapted it to show what I mean: https://gist.github.com/stefan6419846/3d368b26ee5260a7886657909f26ca15 The `adobe_glyphs` module imported there is a standalone copy of https://github.com/py-pdf/pypdf/blob/main/pypdf/_codecs/adobe_glyphs.py

This will generate the code to create instances of the dataclass shown in https://github.com/py-pdf/pypdf/pull/3363#discussion_r2192784169 together with the copyrights. The only things are are currently missing are the dataclass definition itself and the necessary imports as well as the mapping in the footer, but this should be easy enough to add to the script.

By the way: We should probably keep the script outside of the original code, but run some linting and testing on it nevertheless to ensure it matches our standards and does not break for some reason.

PJBrs commented:
@stefan6419846 OK, now I understand better! Basically, you want the script to immediately the specific Font instances, instead of ending up in an intermediate form. With regard to linting / typing, as far as I can tell, I can just run ruff and mypy locally where it can find the script...? At least, that seems to work here.

I understand also, that using this, I can do something like:

```
if font_name in FONT_METRICS:
    font = FONT_METRICS[font_name]
```
and then do stuff like:

```
total_width = sum(font.character_widths[char] for char in "This is a long sentence")
print (total_width)
```

One question; Why do you only add 255 widths? The AFMs contain about 314 widths.

Second; so, this really collects a lot of information in the Font class. In my local patches, I need the widths (and I expect also some other metrics) for text wrapping a text stream when flattening an annotation. With my original patches, I changed build_font_width_map to include the character widths that are now in the Font instances in FONT_METRICS. However, the new Font class would not include character widths that _are_ available for embedded fonts. So, how would you proceed with this? I see two ways forward:
1. Change build_font_width_map and use the new Font instances, something like this:

```
+    else:
+        font_name = str(ft["/BaseFont"])[1:]
+        if font_name in FONT_METRICS:
+            font_width_map = cast(Dict[str, float], FONT_METRICS[font_name].character_widths)

```

2. Change the Font dataclass so that you can initialise it with a /Font resource dictionary, set attributes bases on FONT_METRICS instances when available, and otherwise set character_widths suing the existing build_font_width_map. This makes it possible, in the future, to also set the other metrics based on embedded font information, if available.

stefan6419846 commented:
> With regard to linting / typing, as far as I can tell, I can just run ruff and mypy locally where it can find the script...?

Yes, although for the repository and its CI/CD, we might need to extend the current configuration. I am open to help with this if desired/required.

> One question; Why do you only add 255 widths? The AFMs contain about 314 widths.

I used a mix of the original *pdfminer.six* code and your code for writing my script. The limitation is from *pdfminer.six*. This seems to mostly eliminate the characters with the code `-1`. I have no hard opinion on this/do not know what is correct.

> However, the new Font class would not include character widths that are available for embedded fonts. So, how would you proceed with this?

I would split this into two separate topics. This PR should focus on the new container and retrieving the data from the AFM files to use them where appropriate. Possibly unifying this with handling embedded fonts could/should be a separate step in a PR afterwards. This way, we avoid too large PRs which simplifies reviewing the changes from my side.

PJBrs commented:
@stefan6419846 

I used your script and tried to improve it, and I added your font class.

To be fair, this PR more and more looks like you answering my bug report instead of me trying to contribute new code; please adjust copyright and attribution accordingly if and when you pull this.

I added the font class to pypdf/_font.py. I noticed that other files in this directory have a copyright notice and attribution, please advise, and add your name if you needed!

I just tested locally, and the following works:

```
            font_name = font_res["/BaseFont"]  # [/"Name"] often also exists, but is deprecated
            if font_name[1:] in FONT_METRICS:
                my_font = FONT_METRICS[font_name[1:]]
                print (sum( my_font.character_widths.get(char, 200) for char in "Such a long sentence, how long is€€€𒈙 it " ))

```

I noticed that codecov would like a test. I may be able to cobble up something, and otherwise it will be in a month or so.

stefan6419846 commented on resources/afm_to_dataclass.py:
```diff
@@ -0,0 +1,166 @@
+# ruff: noqa: T201, INP001, D100
+# Use this file to generate Font dataclasses for the 14 Adobe Core fonts:
+# Place it in the pypdf/_codecs directory and run with:
```
We can avoid this with the correct import.

PJBrs commented on resources/afm_to_dataclass.py:
```diff
@@ -0,0 +1,166 @@
+from typing import cast
+from zipfile import ZipFile
+
+from adobe_glyphs import adobe_glyphs
```
Done!

stefan6419846 commented on pypdf/_font.py:
```diff
@@ -0,0 +1,47 @@
+# Copyright (c) Stefan...
```
Similar to `_page_labels.py`, we can leave the header out here (which we do in `afm_to_dataclass.py` as well, where we have much more originality and basically no relevant *pdfminer.six* code anymore). The current class definition is trivial enough.

PJBrs replied:
Done!

stefan6419846 commented on pypdf/_writer.py:
```diff
@@ -797,8 +797,7 @@ def _merge_content_stream_to_page(
         """
-        Combines existing content stream(s) with new content (as bytes),
-        and returns a new single StreamObject.
+        Combines existing content stream(s) with new content (as bytes).
```
This is unrelated to the PR.

PJBrs replied:
Yes, it was a stupid oversight in the previous one. Does it hurt to add as part of this PR? It seems overkill to open a separate PR for this oneliner.

stefan6419846 commented:
> To be fair, this PR more and more looks like you answering my bug report instead of me trying to contribute new code; please adjust copyright and attribution accordingly if and when you pull this.

I am helping with getting the changes integrated. Without you doing the initial work, I would not have looked into this myself. Doing the changes to the parser has been a little side project, allowing me some insights into AFM files.

> I noticed that codecov would like a test. I may be able to cobble up something, and otherwise it will be in a month or so.

Your initial PR integrated the new functionality into the existing code and adapted some tests. Wouldn't this be sufficient or does this have any side effects? Codecov is correctly complaining because the new code is never executed.

PJBrs commented:
> > To be fair, this PR more and more looks like you answering my bug report instead of me trying to contribute new code; please adjust copyright and attribution accordingly if and when you pull this.
> 
> I am helping with getting the changes integrated. Without you doing the initial work, I would not have looked into this myself. Doing the changes to the parser has been a little side project, allowing me some insights into AFM files.
> 
> > I noticed that codecov would like a test. I may be able to cobble up something, and otherwise it will be in a month or so.
> 
> Your initial PR integrated the new functionality into the existing code and adapted some tests. Wouldn't this be sufficient or does this have any side effects? Codecov is correctly complaining because the new code is never executed.

I wrote a very small test file now - tests/test_font.py:
```
"""Test font-related functionality."""


from pypdf._codecs.core_fontmetrics import FONT_METRICS


def test_font_metrics():
    font_name = "Helvetica"
    my_font = FONT_METRICS[font_name]
    assert my_font.family == "Helvetica"
    assert my_font.weight == "Medium"
    assert my_font.ascent == 718
    assert my_font.descent == -207

    test_string = "This is a long sentence. !@%%^€€€. çûįö¶´"
    charwidth = sum(my_font.character_widths[char] for char in test_string)
    assert charwidth == 19251

    font_name = "Courier-Bold"
    my_font = FONT_METRICS[font_name]
    assert my_font.italic_angle == 0
    assert my_font.flags == 64
    assert my_font.bbox == (-113.0, -250.0, 749.0, 801.0)
```

If you prefer, I'll add something to the writer test.

PJBrs commented:
> I am okay with a simple test covering one of the basic entries and possibly one of the explicitly mapped ones as well.

Sounds like what I just pasted, right?

PJBrs commented:
I'm testing some ideas and I get a circular input, I think... Let's let this simmer until I have something that I know works.

PJBrs commented:
So, I tried yesterday to use the new Font class but I ran in quite some trouble, which basically comes down to how to initialise a Font instance. The reason: I would like to initialise a Font using a font resource dictionary.

As it is, everything works quite well for the 14 core fonts. However, it is important to note that those 14 fonts are actually the exception to the rule, the rule being that, for an embedded font, all metrics and its encoding ought to be available as part of a PDF font resource dictionary. Why is this important? Because, in practice, given the above one would want to be able to initialise a Font instance using a font resource dictionary as present within the PDF file.

This leads to an issue of issue of circular imports: the FONT_METRICS dict imports the Font class. However, once you would like to recognise one of these fonts based on a font resource dict at initialisation, the Font class itself would need to import FONT_METRICS. I don’t see how we can work around this. I tried, and I got: 

```
ImportError: cannot import name 'FONT_METRICS' from partially initialized module 'pypdf._codecs.core_fontmetrics' (most likely due to a circular import) (/usr/lib64/python3.9/site-packages/pypdf/_codecs/core_f
ontmetrics.py)
```

From a more integrated perspective on what a font is in PDF terms, the font metrics for the 14 core fonts are incomplete. As a font, they should be accompanied by an encoding and a character map, both of which are specified in the font resource dictionary. I would argue that, what in this PR is a `Font` class, in actuality is more a `FontMetrics` class.

I can think of two ways forward:

1. Turn the current Font class into a FontMetrics class, add that to `_codecs/core_fontmetrics.py,` and if, in the future, pypdf would want to include a full-fledged Font class, then the FontMetrics class would be a subclass in case the user hits a non-embedded instance of a core font.
2. Keep the current (beginnings of a) Font class, turn the current FONT_METRICS back into a Dict[Dict, Any] (or whatever it was, but not a class), and add an __init__ method that takes a font resource dict as argument and uses the FONT_METRICS for the core fonts if those metrics are not available in the font resource dict itself.

I personally would prefer the second option; I see your point that the second option is less object oriented, but I do think that it is more robust for future changes. Nevertheless, please advise!

PJBrs commented:
On second (final?) thought, it occurs to me that probably the first option is best.

A Font dataclass already exists in pypdf/_text_extraction/_layout_mode/_font.py and I it seems that associated code is already  used in _page. Perhaps it is best to just focus on the missing functionality...

PJBrs commented:
I've given all this still more thought. So, the AFM file contains a set of widths and a set of font characteristics that we could call FontDescriptor. Existing pypdf code includes incomplete font width information in two places: https://github.com/py-pdf/pypdf/blob/bfe7178e23244bf5aa875d226b074a75df8ebccc/pypdf/_cmap.py#L107 and https://github.com/py-pdf/pypdf/blob/bfe7178e23244bf5aa875d226b074a75df8ebccc/pypdf/_text_extraction/_layout_mode/_font_widths.py#L2 .

Furthermore code already exists in two places to get a set of font widths from a font resource dictionary, in pypdf/_cmap.py:
https://github.com/py-pdf/pypdf/blob/bfe7178e23244bf5aa875d226b074a75df8ebccc/pypdf/_cmap.py#L402
and in pypdf/_text_extraction/_layout_mode/_font.py:
https://github.com/py-pdf/pypdf/blob/bfe7178e23244bf5aa875d226b074a75df8ebccc/pypdf/_text_extraction/_layout_mode/_font.py#L38

In both cases, an incomplete fallback exists for the 14 core fonts that this PR should be able to solve.

Code dealing with FontDescriptor information is nowhere to be seen yet (as far as I can tell). Such information is being used in pdftk to generate appearance streams. For embedded fonts, this ought to be part of a font resource dictionary. Here's an example of an arbitrary PDF that I opened:
`{'/Type': '/Font', '/Subtype': '/CIDFontType2', '/CIDSystemInfo': {'/Ordering': 'Identity', '/Registry': 'Adobe', '/Supplement': 0}, '/FontDescriptor': IndirectObject(29, 0, 139871477943648), '/BaseFont': '/RIPFSJ+Tahoma,Bold', '/W': [3, [292], 15, [312], 19, [636], 39, [757], 49, [770], 54, [633], 55, [612], 68, [598], 69, [631], 72, [593], 76, [301], 78, [602], 79, [301], 80, [953], 82, [617], 85, [433], 87, [415], 88, [640], 92, [575], 188, [636]]}`

Adding code that parses FontDescriptor information would be trivial. For instance, info like the above is parsed in at least three places already (as far as I can tell). It's just that the FontDescriptor information is not extracted.

This leads me to the following conclusion:

- The Font dataclass in this PR is too broad for the information that it adds
- A FontMetrics dataclass would be better, but the widths part would not be so easy to access and no right place exists yet for the FontDescriptor information

I now think that the most elegant solution would be to:

- Add a FontDescriptor dataclass
- Adapt the afm_to_dataclass.py script so that it creates FONT_METRICS: Dict[str, Tuple[FontDescriptor, Dict[str, int]]] ; in other words, a Dict with font name as key, and a tuple as value. Each tuple consists of a FontDescriptor and a Dict with font widths. This is very similar to the original pdfminer.six code, but with a dataclass for the FontDescriptor instead.

This enables the following subsequent steps (not in this PR):

1. Correct the core font width information in _cmap.py and in _text_extraction/_layout_mode/_font.py ; these are really trivial changes that then make it very ease to adapt the generate_appearance_stream method so that it can correctly wrap, scale, centre and right-align text. <-- This is the big aim that I'm after, and have most of the code for.
2. Create a build_font_descriptor_from_dict method (or maybe add a build_font_descriptor_from_dict classmethod to initialise a FontDescriptor, who knows!)
3. Move the _font.py file to pypdf/_font.py
4. Perhaps add FontDescriptor attribute to Font class

@stefan6419846 , please advise. It's like a menu :-) Would you like:

1. The PR as it currently is;
2. Turn the current Font class into a FontMetrics class, add that to _codecs/core_fontmetrics.py, as discussed above
3. Keep the current (beginnings of a) Font class, turn the current FONT_METRICS back into a Dict[Dict, Any] (or whatever it was, but not a class), and add an init method that takes a font resource dict as argument (I now think this is a bad idea)
4. The above proposal. (This I like best).

I'm going to redo the PR in about three weeks or so.

stefan6419846 commented on pypdf/_font.py:
```diff
@@ -0,0 +1,33 @@
+    family: str = "Unknown"
+    weight: str = "Unknown"
+
+    ascent: float = 800.0  # A typical ascent value for many fonts
```
Where are these defaults coming from?

PJBrs commented:
@stefan6419846 Thanks for your previous review and especially for asking about the defaults. It's clear to me now that most defaults offered by Google Gemini were quite alright, but the whole flags thing was very probably wrong. I should have been more critical myself.

Interestingly, the original code that set the font flags to 64 for Courier (as copied from pdfminer.six) appears wrong as well; it tests for the IsFixed value, but that corresponds to the first bit of the flags (in other words, 1, not 64).

I don't think the flags are really important, otherwise the people at pdfminer.six would have noticed. I do think that I now have the correct values, having looked at Section 9.8.2 Font descriptor flags of the pdf specification.

PJBrs commented:
Enjoy your travels!
